### ISSUE TITLE
Survey on VueJS vs ReactJS and Tripal (& extensions)

### ISSUE THREAD

**Issue by guignonv** [tripal/tripal#1045]
This is more a Tripal community question rather than a core Tripal question.

**Context**:
Drupal 9 should be released in 2020 and Drupal 7 will be end-of-life in November 2021. The  community is currently working on Tripal 4 for Drupal 8 which *should* be easily ported to Drupal 9 as well (minor work expected). It means that you will soon be strongly encouraged to move to Drupal 8/9 and setup your new website. It might be an opportunity to update your theme and maybe move to "newer" technologies.
 
For instance, you may want to use a modern front-end/ single page application (SPA) framework. However, there are many front-end JS frameworks to choose from.  We propose that our community would benefit from adopting one as the recommended Tripal framework.  This would improve collaboration and may prevent compatibility issues.

**More context** (SPA/front-end frameworks):
In that jungle, we consider the two most popular frameworks, VueJS (https://vuejs.org/) and ReactJS (https://reactjs.org/). We recognize the main considerations for the choice are: learning curve, efficiency, sustainability, and Drupal compatability.
 
VueJS is considered by many to be the easiest to learn. However, it seems that the Drupal community will choose ReactJS (https://www.drupal.org/project/ideas/issues/2913628#comment-13470993) as they already started to use it for the admin user interface (https://www.drupal.org/about/strategic-initiatives/admin-ui-js).

**Sum-up pros/cons**:
Vue:
- supposed to provide the best features of Angular and React
- easy to learn compared to others: may be suitable for interns or people who can't spend much time on learning such a technology.

React:
- appears to be the future Drupal community choice
- harder to learn

**Other decision elements**:
Both can and have been integrated to Drupal 8 (https://reactfordrupal.com/, https://drupalize.me/series/drupal-8-and-reactjs, https://opensenselabs.com/blog/tech/how-build-single-page-application-drupal-8-and-vue-js). It is possible to use VueJS even if ReactJS is also used by Drupal. Both can be integrated with the Bootstrap theme (https://bootstrap-vue.js.org/docs, https://www.pluralsight.com/guides/how-to-integrate-bootstrap-with-reactjs). Both have are widely adopted with a large community and good documentation.

**The question**:
Which SPA framework would you prefer to adopt (or have you already adopted): VueJS or ReactJS?
For ReactJS react to this post with the flag "Rocket" and for VueJS use the "Eyes".
If you have no opinion, use the "Heart" and if you have another choice use the "Confused" face.  As always we would welcome opinions and questions here.

PS: Thanks Bradford for the review and Abdullah for some inputs.

**Comment by risharde** [comment]
This is a tough one and I must admit I have not used both of these frameworks really out of fear that I would have lack of control in my development or slow down my development speed (I'm kind of impatient when something takes months to code to see a result). With that being said I don't have full confidence in some of the decisions made to Drupal 8/9 - example: even composer installs can be a pain at the moment (memory errors even with up to 2GB of memory for a clean install - yet Drupal recommends this as the preferred method), up to now I can't seem to find a decent admin ui plugin/module for module installing (reference to Wordpress built in plugin finder and installer - and no this doesn't mean I like wordpress lol) and lastly changing how drush works (this might not be Drupal core community) and now using a more complicated framework? (ReactJS). From my googling and youtubing a lot of people say the same as above, that VUE is easier. If the argument is that React is somehow faster, what sense does this have if Drupal backend stuff is slow (still?!) - I hope you guys understand my point through these examples. 

Unfortunately, once again, we may be forced into having to use React since we will probably have to learn it for something else in Drupal if as you said, it's going to be used for the admin. So once again, Drupal team wins and once again, UNLESS we can somehow implement VUE better - but again, to keep also skill viable in an ever changing workplace, we could still be forced to learn REACT.

TLDR;
My suggestion: Use the React based on the thought process :(. Though I would go with whatever the Tripal core decided to go with of course.

**Comment by spficklin** [comment]
I have not used either framework before but my vote is to go with the one that gives us the best support. If the Drupal community is going with ReactJS and that is where the support focus will be then I think we would be better off in the long-term sticking with the crowd.  

But I agree with @risharde regarding his thoughts on composer. I'm not a fan of it.

**Comment by almasaeed2010** [comment]
We use both frameworks in the Staton lab to develop applications outside of Drupal. I haven't used either with Drupal yet. 

Although I do like the simplicity of VueJS, I think we should stick with what Drupal supports out of the box mainly because it doesn't make sense to me to load 2 major JS libraries into a single application. Integrating VueJS components into ReactJS and vice versa is possible but not trivial so I'd recommend against it.

**Comment by ghost** [comment]
I'm in the design phase for a new website to include a new database to identify patterns of resistance and the mechanisms that allow HIV-1 strains to escape neutralization from bNAbs. The university will no longer allow new sites to be setup as Drupal 7. Can someone advise me on the timing of Tripal 4 on Drupal 8? Further, not having developed with Tripal yet, is Tripal the right direction for a Drupal site which will store HIV-1 sequences and bNAbs in the database?  I was planning to be up on Tripal using JBrowse for sequence viewing and then continuing with a JS framework for analysis tools.

**Comment by laceysanderson** [comment]
Hi @pst024 and welcome to the community! We are currently working on Tripal 4 for Drupal 8/9 in this repository for organizational reasons: https://github.com/tripal/t4d8. As you can see at the bottom of the Readme and added here for ease, we've developed a timeline for development to ensure existing Tripal sites have time to upgrade before the D7 end of life.
[<img width="1206" alt="Screen Shot 2020-11-25 at 9 13 32 AM" src="https://user-images.githubusercontent.com/1566301/100246243-84d42a80-2efe-11eb-966c-ebe88e899cb6.png">](https://time.graphics/line/387503)

There is already a lot the core functionality done and we are just putting the finishing touches on the Fields functionality and API for a release before the end of the year! You could definitely start developing for Drupal 8.8+ now and then add in content pages using Tripal once it's released. 

And yes, Tripal is definitely the right direction for any biological data stored in a Drupal site! It provides functionality for managing biological data through Drupal including an ontology-focused approach, a specialized content entity type and fields integrated tightly with Drupal and semantic web ready web services for all Tripal content. It  also provides support for long running command-line focused jobs, biological data visualization and integration with external applications. 

Regarding JBrowse, we have an extension module specific to integrating JBrowse1 and soon JBrowse2 into a Drupal/Tripal site here: https://github.com/tripal/tripal_jbrowse. As a core developer of that module I can guarantee it will be one of the first modules upgraded to Drupal 8/9.

**Comment by ghost** [comment]
Awesome! Thank you the update and quick response! I'm excited to use Tripal!

**Comment by spficklin** [comment]
It's been awhile since we chatted on this. @guignonv . I'm curious if you have any thoughts more on this as months have passed?  Have you tried any more work on this?

**Comment by guignonv** [comment]
@spficklin not much thoughts yet. I'm -almost!- starting the new implementation of MGIS on Drupal 9 with more than a year late (at the moment) and I'll have to try those technologies and decide. I keep you tuned. ;-)
